File: Java/LECCION5/Encapsulamiento/src/dominio/Persona.java
/*
podemos ver el estado del objeto,tenemos varis datos, la mejor practica es encapsular varios atributos
ahota vamos autilizar un modificador de acceso llamdo private,este evita el acceso directo,para acceder a los datos 
solo sera por medio de otros metodos, esos metodosvan a ser el GETTER Y SETTER,elel estado del objeto es que va a estar encapsulado 
y la unica manera para acceder es atravez de lo externoque son los metodos de los objetos, inertface o metodos del objeto
son los get y los set y es la unica manera para poder llegar a los datos, el estado del objeto
como podemos observar la idea de encapsulamientode la programacion orientada a objetos es evitar el acceso y manipulacion directa de otras clases
a los datos de un objeto por datos nos referimos a cuaslquier atributo de la clase en cuestion,esto es el estado del objeto,para acceder a los
datos solo sera por medio de los metodos setter y getter
-Recuerden que Setter modifica el valor del atributo
-y los Getter obtienen informacion del atributo ,los getter y los setter van a ser los metodos de acceso y estos si deben ser publicos  para que logremos
acceder y los metodos setter y getter debemos crear uno por cada atributo 
*/
package dominio;


public class Persona { 
    /*ATRIBUTOS ESTA CLASE LA VAMOS ACREAR CON ATRIBUTOS DE TIPO PRIVATE
    que pasa si ponemos private con MAYUSCULA, no lo toma, es muy importante para Java lo que son MAYUSCULAS Y MINUSCULAS
    el tipo STRING comienza con Mayuscula porque String no esta dentro de lo que son los tipos primitivos 
    los tipos primitivos son double,boolean,int,long,todos estos son tipos primitivos, String no entra dentro de los grupos primitivos
    por eso debe comenzar con Mayusculas a diferencia de los otros
    los atributos se han marcado de color verde (nombre,sueldo y eliminado)las palabras reservadas el editor las muestra en azul
    */
    private String nombre;
    private double sueldo;
    private boolean eliminado;
    /*VAMOS A CREAR AL CONSTRUCTOR de la clase Persona que va hacer de tipo Publico
    recuerden que siempre el CONSTRUCTOR debe llevar el mismo nombre que la clase, pero ya si nuestros atributosa son de 
    tipo privateya no se pueden acceder a ellos, la unica manera de acceder a ellos es atravez de los metodos 
    en este caso EL METODO PRINCIPAL ES EL CONSTRUCTOR, entonces el CONSTRUCTOR debe ser de tipo PUBLICO si no o podriamos acceder a los atributos
    dentro de los parentesis necesitamos definir lo que serian los atributos,las variables que vamos a recibir para modificar estos atributos 
    */
  public Persona(String nombre, double sueldo, boolean eliminado){ //pasamos las variables que serian nombre,sueldo y elim...
      this.nombre = nombre; //vamos a usar this nombre se va aponer de verde mas oscuro este es el atributo el otro nombre es la variable
      this.sueldo = sueldo; // de la linea 34 a la 38 tenemos creado el contructor
      this.eliminado = eliminado;
    }
  //VAMOS A CREAR LOS METODOS QUE NECESITAMOS Y estos metodos van a ser el METODO SETTER Y GETTER 
  public String getNombre(){
      return nombre;
  }
  public void setNombre(String nombre){// es el que modifica 
      this.nombre =nombre;
      
  }
  public double getSueldo(){
      return sueldo;
  }
  public void setSueldo(double sueldo){ //los get no devuelven nada pero si necesitan recibir un parametro, que es el que va a modificar nuestro atributo
      this.sueldo =sueldo;
      
  }
  public boolean isEliminado(){ //"aca en get nombre vamos a USAR EL IS porque los booleanos hace una pregunta" solo cambia para el get en los boolean
      return eliminado;
  }
  public void setElimidado(boolean eliminado){
      this.eliminado = eliminado;
  }
    
}

File: Python/LECCION7/Persona.py
# si quisieramos poner que la clase Persona esta heredando de la clase object se abren los parentesis y se pone object
# de esa manera estaria heredando la clase Persona de la clase object, pro no hace falta poner dentro del parentesis
#esto se hace automaticamente

class Persona():# luego ponemos el metodo inicializador para nuestros atributos el metodo dander init/ Esta clase hereda de la clase object
    def __init__(self, nombre, edad): #aqui necesitamos pasar los parametros
        self._nombre = nombre
        self._edad = edad

    @property
    def nombre(self):
        return self._nombre

    @nombre.setter
    def nombre(self,nombre):
        self._nombre = nombre

    @property
    def edad(self):
        return self._edad

    @edad.setter
    def edad(self, edad):
        self._edad = edad
    def __str__(self): # Override = sobreescribir los atributos todo lo que recibe al ser miembro de la misma clase es que puede acceder a los metodos encapsulados,no los muestra con el valor real
        return  f'Persona:  [Nombre:{self._nombre}, Edad: {self._edad} ]'# este metodo nos va a mostrar de tipo cadena tipo string, este metodo tiene un return y usaremos la interpolacion pero desde afuera
#en la linea 25 nos muestra el overrrides method in object, la linea 25 y 26 sobrescribir quiere decir que un metodo definido en una clase Padre lo volvemos a definir en una clase hija ,esto es escribir
#o representar lo echo en la clase hija es la manera de describir el comportamiento y de definir n metodo de la clase padre ahora en la clase hija ,tambien es el metodo init el que se sobresribe en la clase hija
#lo hemos hechoe en la linea 34 para poder acceder al atributo de lo que es empleado ahora aqui de lo que es cliente hemos accedido a los atributos de lo que es la clase persona, pero no lo que es a la clase emplea
#do clase hija


class Empleado(Persona):    #aqui vamos hacer la herencia de la clase Persona, de esta forma decimos que la clase empleado es hija de la clase Persona
    def __init__(self,nombre,edad, sueldo): #El parametro que necesita es sueldo
        super().__init__(nombre , edad) # es el contructor de la clase Padre de todos object
        self._sueldo = sueldo

    @property
    def sueldo(self):
        return self._sueldo

    @sueldo.setter
    def sueldo(self, sueldo):
        self._sueldo = sueldo

    #Aqui vamos a poner el metodo dander str

    def __str__(self):
        return f'Empleado: [ sueldo: {self._sueldo} ]{super().__str__()}'


# TAREA HACER CON LOS METODOS GETTER Y SETTER LOS ENCAPSULAMIENTOS
 #TAREA : ENCAPSULAR LOS ATRIBUTOS Y AGREGAR LOS METODOS GETTERS AND SETTERS
 #CREAR OTRO OBJETO PASAR LOS DATOS PARA NOMBRE EDAD Y SUELDO
#saliendo de las clases vamos a ver si podemos crear un objeto
# MOSTRAR ESTOS DATOS ,LUEGO MODIFICAR Y MOSTRAR NUEVAMENTE

empleado1 = Empleado('Ariel', 40, 75000)
print(empleado1.nombre)
print(empleado1.edad)
print(empleado1.sueldo)

empleado2 = Empleado('Liliana', 38, 70000)
print(empleado2.nombre)
print(empleado2.edad)
print(empleado2.sueldo)
empleado2.nombre = 'Natalia'
empleado2.edad = 35
empleado2.sueldo = 75000
print(empleado2.nombre)
print(empleado2.edad)
print(empleado2.sueldo)



